### DISCUSSION TITLE
Variable beta?

### DISCUSSION THREAD

**Discussion by mncrowe** [FourierFlows/GeophysicalFlows.jl#368] [Ideas]
From a technical rather than physical perspective, it might be useful to do variable beta in simulations (e.g. for parameter continuation). I'm guessing this isn't currently possible as L depends on beta and is immutable? Is there currently a way to do this and if not, perhaps it could be considered as an addition?

Matthew

**Comment by glwagner** [comment]
Huh, isn't `L` an array? In that case it is mutable, you can simply update `L` with a new beta:

https://github.com/FourierFlows/GeophysicalFlows.jl/blob/1dd98e4b9d7e6fe421d1423db8bad70fe646d07f/src/singlelayerqg.jl#L231

**Reply by mncrowe** [reply]
`ERROR: setfield!: immutable struct of type Equation cannot be changed`

The issue is L is buried inside the Equation structure from FourierFlows.jl

**Comment by glwagner** [comment]
There are two other ways you could achieve this:

## `with_params` utility

```julia
function with_params(problem, new_params)
  new_equation = Equation(new_params, problem.grid)
  return FourierFlows.Problem(new_equation, problem.stepper, problem.dt, problem.grid, problem.vars, new_params)
end
```

That's fast and easy, you re-use the memory allocated for the old problem.

## change_params! utility

You could put in a PR to change `FourierFlows.Problem` to be mutable. There's no reason why it shouldn't be (it's not even a concrete type at the moment):

https://github.com/FourierFlows/FourierFlows.jl/blob/1c412b9708fd5a22a429b43dc51b50b6e5dd7bba/src/problem.jl#L59

Then you could design utility that allows you to change the parameters, eg

```julia
function change_params!(problem, new_params)
  new_equation = Equation(new_params, problem.grid)
  problem.equation = new_equation
  problem.params = new_params
  return nothing
end
```

**Reply by mncrowe** [reply]
Thanks, `with_params` is a nice way to do it. I'll try that for now, might do the PR later depending on how much I need to do this.

**Reply by glwagner** [reply]
You should contribute `with_params` either way! If its useful to you it will probably be useful to others
